### PR TITLE
internal: Replaced adding vo to rse_expression with filters, #51

### PR DESCRIPTION
**Pull request by patrick-austin**
internal: Replaced adding vo to rse_expression with filters, #51
------------------
See issue #51 for description of changes.

### CODE REVIEW

File: lib/rucio/core/replica.py [hunk 1]
```diff
@@ -742,7 +742,7 @@ def _list_replicas_for_files(file_clause, state_clause, files, rse_clause, updat
 def _list_replicas(dataset_clause, file_clause, state_clause, show_pfns,
                    schemes, files, rse_clause, rse_expression, client_location, domain,
                    sign_urls, signature_lifetime, constituents, resolve_parents,
-                   updated_after, vo,
+                   updated_after, filters,
```
Comment (elichad) [inline lib/rucio/core/replica.py:745]: Why is this argument `filters` and not `filter`?

File: lib/rucio/core/rse.py [hunk 2]
```diff
@@ -367,7 +368,19 @@ def list_rses(filters={}, session=None):
     availability_mask2 = 7
     availability_mapping = {'availability_read': 4, 'availability_write': 2, 'availability_delete': 1}
     false_value = False  # To make pep8 checker happy ...
+
     if filters:
+        filters_to_process = len(filters)
```
Comment (elichad) [inline lib/rucio/core/rse.py]: I don't understand why this `filters_to_process` stuff is needed. Could you explain?

File: lib/rucio/core/rse_expression_parser.py [hunk 3]
```diff
@@ -92,26 +92,37 @@ def parse_expression(expression, filter=None, session=None):
...
-        # Don't cache the result if we only specified vo
-        vo_pattern = r'^vo=[a-zA-Z0-9]+$'
-        vo_match = re.search(vo_pattern, expression)
-        if vo_match is None:
-            REGION.set(sha256(expression.encode()).hexdigest(), result)
+        REGION.set(sha256(expression.encode()).hexdigest(), result)
 
-    if not result:
+    if filter:
+        filters_to_process = len(filter)
```
Comment (elichad) [inline lib/rucio/core/rse_expression_parser.py]: Same as above

File: lib/rucio/daemons/bb8/common.py [hunk 4]
```diff
@@ -83,19 +84,11 @@ def rebalance_rule(parent_rule, activity, rse_expression, priority, source_repli
...
-        rse_expression = 'vo={}&({})'.format(rule_vo, rse_expression)
-
     # check if concurrent replica at target rse does not exist
     concurrent_replica = False
     try:
         for lock in get_dataset_locks(parent_rule['scope'], parent_rule['name']):
             lock_rse_expr = lock['rse']
-            if rule_vo != 'def':
-                lock_rse_expr = 'vo={}&({})'.format(rule_vo, lock_rse_expr)
             if lock_rse_expr == rse_expression:  # may need to evaluate to be sure... could get 'vo=tst&(vo=tst&(MOCK))'
```
Comment (elichad) [inline lib/rucio/daemons/bb8/common.py]: I think this comment can be removed since this should no longer be a concern

**Review by elichad — CHANGES_REQUESTED** [review]

File: lib/rucio/core/replica.py (hunk 1 above)
Comment (patrick-austin) [reply lib/rucio/core/replica.py:745]: At line 767/766 we actually use the built in python `filter()` function, so the normal practice of naming it `filter` leads to an overwrite and the whole function falls over (like the default arguments for `{}` this another one that pylint doesn't like but normally you can get away with). There are other places we use the argument `filters`, so I just used that instead.

Comment (patrick-austin) [reply lib/rucio/core/replica.py:745]: In general and from now on we should use the term `filters` but it's not worth retroactively changing `filter` to `filters`.

File: lib/rucio/core/rse.py (hunk 2 above)
Comment (patrick-austin) [reply lib/rucio/core/rse.py]: Following discussion with Martin implement a nicer version of this (copying the dict and popping) that doesn't alter the existing filter logic. There may be an underlying issue with the way availability filters work but don't alter that until Martin's had a further look.

**Review by elichad — APPROVED** [review]

**Pull request merged**